ledger/src/blockstore.rs

```diff
@@ -2712,19 +2712,48 @@ impl Blockstore {
     }
 
     pub fn get_recent_perf_samples(&self, num: usize) -> Result<Vec<(Slot, PerfSample)>> {
-        Ok(self
+        // When reading `PerfSamples` the database may contains samples encoded as both
+        // `PerfSampleV1` and `PerfSampleV2`.  We expect `PerfSampleV1` to be a prefix of the
+        // `PerfSampleV2` encoding (see [`perf_sample_v1_is_prefix_of_perf_sample_v2`]), so we try
+        // them in order.
+        let samples = self
             .db
             .iter::<cf::PerfSamples>(IteratorMode::End)?
             .take(num)
             .map(|(slot, data)| {
-                let perf_sample = deserialize(&data).unwrap();
-                (slot, perf_sample)
-            })
-            .collect())
+                // Clippy from 2022-11-02 nightly does not seem to understand that `block 'v2` can
+                // only work inside a labelled block.
+                #[allow(clippy::unnecessary_operation)]
+                'v2: {
+                    match deserialize::<PerfSampleV2>(&data) {
+                        Ok(sample) => return Ok((slot, sample.into())),
+                        Err(err) => {
+                            if let bincode::ErrorKind::Io(io_err) = &*err {
+                                if matches!(io_err.kind(), ErrorKind::UnexpectedEof) {
+                                    break 'v2;
+                                }
+                            }
+                            return Err(err.into());
+                        }
+                    }
+                };
```

Comment:
I don't find this named-block syntax better/less confusing than nesting v1, eg:
```suggestion
                deserialize::<PerfSampleV2>(&data)
                    .map(|sample| (slot, sample.into()))
                    .or_else(|err| {
                        if let bincode::ErrorKind::Io(io_err) = &*err {
                            if matches!(io_err.kind(), ErrorKind::UnexpectedEof) {
                                return deserialize::<PerfSampleV1>(&data)
                                    .map(|sample| (slot, sample.into))
                                    .map_err(|err| err.into());
                            }
                        }
                        Err(err.into())
                    }
                }
```
(Also I think we can get more concise and rust-y than matching.)
Change my mind?

Reply:
I was trying to separate code that parses different versions.
With chaining, it looks like it is more connected.  And is a bit harder to debug.
Plus, when version are split, when v3 is added, it would look similar to v2, not touching any existing code.

But, I must admit, I am not super happy with my version.
In particular, it is annoying that I need a block in order to introduce a break label.
Adds syntactic noise.

And multiple `return` statements add a lot of needless complexity as well.
Could not find a way to avoid them, as `io::Error` is not an enum, and requires a `.kind()` call.

Used your version, though I kept the error type as `Box<bincode::ErrorKind>` until the end of the chain.
This way, v3 addition should be more straightforward.
Inverted the error kind check trying to make the `deserialize()` call a bit more visible.

Reply:
Awesome, better than my suggestion. Thanks!